crates/mpz-zk/src/verifier.rs

```
@@ -0,0 +1,281 @@
+            let RCOTSenderOutput {
+                keys: svole_keys, ..
+            } = self.ot.try_send_rcot(128).map_err(Error::ot)?;
+
+            let uv = ctx.io_mut().expect_next().await?;
+            verifier.check(&svole_keys, uv)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<OT> Vm<Binary> for Verifier<OT>
+where
+    OT: RCOTSender<Block>,
+{
+    type Error = Error;
+
+    fn call_raw(&mut self, call: Call) -> Result<Slice> {
+        let output = self.store.alloc_output(call.circ().output_len());
+
+        let mut count = call.circ().and_count();
+
+        // If the callstack is empty, we allocate more for the consistency check.
+        if self.callstack.is_empty() {
+            count += 128
+        }
+
+        self.ot.alloc(count).map_err(Error::ot)?;
+        self.callstack.push((call, output));
+
+        Ok(output)
+    }
+}
+
+impl<OT> Memory<Binary> for Verifier<OT>
+where
+    OT: RCOTSender<Block>,
+{
+    type Error = Error;
+
+    fn alloc_raw(&mut self, size: usize) -> Result<Slice> {
+        self.store.alloc_raw(size).map_err(Error::from)
+    }
+
+    fn assign_raw(&mut self, slice: Slice, data: BitVec) -> Result<()> {
+        self.store.assign_raw(slice, data).map_err(Error::from)
+    }
+
+    fn commit_raw(&mut self, slice: Slice) -> Result<()> {
+        self.store.commit_raw(slice).map_err(Error::from)
+    }
+
+    fn get_raw(&self, slice: Slice) -> Result<Option<BitVec>> {
+        self.store.get_raw(slice).map_err(Error::from)
+    }
+
+    fn decode_raw(&mut self, slice: Slice) -> Result<DecodeFuture<BitVec>> {
+        self.store.decode_raw(slice).map_err(Error::from)
+    }
+}
+
+impl<OT> View<Binary> for Verifier<OT>
+where
+    OT: RCOTSender<Block>,
+{
+    type Error = Error;
+
+    fn mark_public_raw(&mut self, slice: Slice) -> Result<()> {
+        self.store.mark_public_raw(slice).map_err(Error::from)
+    }
+
+    fn mark_private_raw(&mut self, slice: Slice) -> Result<()> {
```

this should return an error since there are no private values for a zk verifier

---

https://github.com/privacy-scaling-explorations/mpz/pull/261